md/options.md
# Options


To adjust these options, edit:

    ~/.config/chat-script/chat-script.ini

## App

| Option        | Desc                                                                               | Default       |
| ------------- | ---------------------------------------------------------------------------------- | ------------- |
| share         | Whether to create a publicly shareable link for the gradio app                     | False         |
| server_name   | IP address that local app is deployed at                                           | 127.0.0.1     |
| server_port   | Port that local app is deployed at                                                 | 7860          |
| inbrowser     | Whether to automatically launch the gradio app in a new tab on the default browser | True          |

## Chain

| Option               | Desc          | Default                |
| -------------------- | ------------- | ---------------------- |
| embeddings_model     | Name of Ollama LLM used to generate embeddings                                                                                     | mxbai-embed-large      |
| chat_model           | Name of Ollama LLM used to generate responses                                                                                      | mistral                |
| moderation_model     | Name of Ollama LLM used to moderate queries                                                                                        | xe/llamaguard3         |
| chat_url             | URL of Ollama LLM used to generate responses                                                                                       | http://localhost:11434 |
| moderation_url       | URL of Ollama LLM used to moderate queries                                                                                         | http://localhost:11434 |
| show_progress        | Whether to display embeddings model batch progress                                                                                 | False                  |
| keep_alive           | How long the model will stay loaded into memory                                                                                    | 5m                     |
| temperature          | The temperature of the chat model. Increasing the temperature will make the model answer more creatively                           | 0.6                    |
| top-k | Reduces the probability of generating nonsense. A higher value will give more diverse answers, while a lower value will be more conservative      | 30                     |
| top-p                | Works together with top-k. A higher value will lead to more diverse text, while a lower value will generate more conservative text | 0.7                    |
| collection_name      | Name of local document collection                                                                                                  | rag-chroma             |
| top_n_results        | Amount of documents to return                                                                                                      | 3                      |
| rag_fusion           | Whether to enable rag-fusion, an advanced rag technique that may improve semantic search relevance                                 | True                   |
| num_queries          | Number of synthetic queries to generate for rag-fusion                                                                             | 2                      |
| top_n_results_fusion | Maximum amount of documents to return for rag-fusion (maximum, as unique union is taken)                                           | 2                      |
| embeddings_gpu       | Whether to use the GPU when generating embeddings (on devices with <8GB VRAM, setting to False can reduce latency)                 | True                   |

## Embeddings

| Option             | Desc                                                                    | Default           |
| ------------------ | ----------------------------------------------------------------------- | ----------------- |
| embeddings_model   | Name of Ollama LLM used to generate embeddings                          | mxbai-embed-large |
| show_progress      | Whether to display document loading and embeddings model batch progress | True              |
| collection_name    | Name of local document collection                                       | rag-chroma        |
| use_multithreading | Whether to enable CPU multithreading for loading documents              | True              |
| chunk_size         | Number of tokens in each split document chunk                           | 250               |
| chunk_overlap      | Number of tokens shared between consecutive split document chunks       | 50                |
| batch_size         | Maximum number of split documents in each embeddings batch              | 41666             |

## Response

| Option               | Desc                                                                                                                                                                      | Default       |
| -------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------- |
| context_stream_delay | Amount of time in s to delay tokens streamed for non-LLM text (sources, moderation notice)                                                                                | 0.075         |
| max_history          | Maximum number of previous user messages to include as context                                                                                                            | 2             |
| print_state          | Whether to print app state for each query. Includes: IP address, chat history, and context                                                                                | True          |
| moderate             | Whether to moderate user queries before allowing responses. Prints IP address and offending query even if print_state is false (allows for privacy-preserving moderation) | False         |
| moderate_alert       | Whether to display system alerts when an unsafe question is received (Linux-only)                                                                                        | False         |


src/options.py
"""Creates and reads options at ~/.config/chat-script/chat-script.ini"""

import os
from configparser import ConfigParser

CONFIG_DIR = "~/.config/chat-script"
CONFIG_FILE = f"{CONFIG_DIR}/chat-script.ini"

def create():
    """Create options file at ~/.config/chat-script/chat-script.ini with defaults"""
    configuration = ConfigParser()
    configuration['APP'] = {
        'share': 'False',
        'server_name': '127.0.0.1',
        'server_port': '7860',
        'inbrowser': 'True'
    }
    configuration['CHAIN'] = {
        'embeddings_model': 'mxbai-embed-large',
        'chat_model': 'mistral',
        'moderation_model': 'xe/llamaguard3',
        'chat_url': 'http://localhost:11434',
        'moderation_url': 'http://localhost:11434',
        'show_progress': 'False',
        'keep_alive': '5m',
        'temperature': '0.6',
        'top_k': '30',
        'top_p': '0.7',
        'collection_name': 'rag-chroma',
        'top_n_results': '3',
        'rag_fusion': 'True',
        'num_queries': '2',
        'top_n_results_fusion': '2',
        'embeddings_gpu': 'True'
    }
    configuration['EMBEDDINGS'] = {
        'embeddings_model': 'mxbai-embed-large',
        'show_progress': 'True',
        'collection_name': 'rag-chroma',
        'use_multithreading': 'True',
        'chunk_size': '250',
        'chunk_overlap': '50',
        'batch_size': '41666'
    }
    configuration['RESPONSE'] = {
        'context_stream_delay': '0.075',
        'max_history': '2',
        'print_state': 'True',
        'moderate': 'False',
        'moderate_alert': 'False'
    }
    with open(os.path.expanduser(CONFIG_FILE), 'w', encoding='UTF-8') as configfile:
        configuration.write(configfile)
    print(f"\nCreated CONFIG_FILE at: {CONFIG_FILE} and populated it with default settings")

def read():
    """Read options from ~/.config/chat-script/chat-script.ini and save in global dict: options"""
    configuration = ConfigParser()
    configuration.read(os.path.expanduser(CONFIG_FILE))
    global OPTIONS
    OPTIONS = {
        'app': {
            'share': configuration.getboolean(
                "APP",
                "share",
                fallback=False
            ),
            'server_name': configuration.get(
                "APP",
                "server_name",
                fallback="127.0.0.1"
            ),
            'server_port': configuration.getint(
                "APP",
                "server_port",
                fallback=7860
            ),
            'inbrowser': configuration.getboolean(
                "APP",
                "inbrowser",
                fallback=True
            )
        },
        'chain': {
            'embeddings_model': configuration.get(
                "CHAIN",
                "embeddings_model",
                fallback="mxbai-embed-large"
            ),
            'chat_model': configuration.get(
                "CHAIN",
                "chat_model",
                fallback="mistral"
            ),
            'moderation_model': configuration.get(
                "CHAIN",
                "moderation_model",
                fallback="xe/llamaguard3"
            ),
            'chat_url': configuration.get(
                "CHAIN",
                "chat_url",
                fallback="http://localhost:11434"
            ),
            'moderation_url': configuration.get(
                "CHAIN",
                "moderation_url",
                fallback="http://localhost:11434"
            ),
            'show_progress': configuration.getboolean(
                "CHAIN",
                "show_progress",
                fallback=False
            ),
            'keep_alive': configuration.get(
                "CHAIN",
                "keep_alive",
                fallback="5m"
            ),
            'temperature': configuration.getfloat(
                "CHAIN",
                "temperature",
                fallback=0.6
            ),
            'top_k':  configuration.getint(
                "CHAIN",
                "top_k",
                fallback=30
            ),
            'top_p': configuration.getfloat(
                "CHAIN",
                "top_p",
                fallback=0.7
            ),
            'collection_name': configuration.get(
                "CHAIN",
                "collection_name",
                fallback="rag-chroma"
            ),
            'top_n_results': configuration.getint(
                "CHAIN",
                "top_n_results",
                fallback=3
            ),
            'rag_fusion': configuration.getboolean(
                "CHAIN",
                "rag_fusion",
                fallback=True
            ),
            'num_queries': configuration.getint(
                "CHAIN",
                "num_queries",
                fallback=2
            ),
            'top_n_results_fusion': configuration.getint(
                "CHAIN",
                "top_n_results_fusion",
                fallback=2
            ),
            'embeddings_gpu': configuration.getboolean(
                "CHAIN",
                "embeddings_gpu",
                fallback=True
            )
        },
        'embeddings': {
            'embeddings_model': configuration.get(
                "EMBEDDINGS",
                "embeddings_model",
                fallback="mxbai-embed-large"
            ),
            'show_progress': configuration.getboolean(
                "EMBEDDINGS",
                "show_progress",
                fallback=True
            ),
            'collection_name': configuration.get(
                "EMBEDDINGS",
                "collection_name",
                fallback="rag-chroma"
            ),
            'use_multithreading': configuration.getboolean(
                "EMBEDDINGS",
                "use_multithreading",
                fallback=True
            ),
            'chunk_size': configuration.getint(
                "EMBEDDINGS",
                "chunk_size",
                fallback=250
            ),
            'chunk_overlap': configuration.getint(
                "EMBEDDINGS",
                "chunk_overlap",
                fallback=50
            ),
            'batch_size': configuration.getint(
                "EMBEDDINGS",
                "batch_size",
                fallback=41666
            )
        },
        'response': {
            'context_stream_delay': configuration.getfloat(
                "RESPONSE",
                "context_stream_delay",
                fallback=0.075
            ),
            'max_history': configuration.getint(
                "RESPONSE",
                "max_history",
                fallback=2
            ),
            'print_state': configuration.getboolean(
                "RESPONSE",
                "print_state",
                fallback=True
            ),
            'moderate': configuration.getboolean(
                "RESPONSE",
                "moderate",
                fallback=False
            ),
            'moderate_alert': configuration.getboolean(
                "RESPONSE",
                "moderate_alert",
                fallback=False
            )
        }
    }
